File: packages/domain/src/query-packs/pack.ts
/**
 * Query pack construction, canonical hashing, and version identity (BB-038).
 */
import { hashUtf8 } from '../provenance/hashes.js';
import { assertQueryTermsValid } from './terms.js';
import type {
  QueryPack,
  QueryPackFixture,
  QueryPackVersion,
  QueryPackVersionId,
  QueryTerm,
} from './types.js';
import { QUERY_PACK_SCHEMA_VERSION } from './types.js';
import type { EntityKind } from '../entity-kinds.js';
import type { QueryPackTheme } from './types.js';

const SEMVER_PATTERN = /^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$/;

export function assertSemverValid(semver: string): void {
  if (!SEMVER_PATTERN.test(semver.trim())) {
    throw new Error(`Invalid semver for query pack: ${semver}`);
  }
}

type CanonicalPackContent = {
  readonly id: string;
  readonly displayName: string;
  readonly entityKind: EntityKind;
  readonly theme: QueryPackTheme;
  readonly semver: string;
  readonly terms: readonly QueryTerm[];
  readonly notes?: string;
};

function canonicalizeTerms(terms: readonly QueryTerm[]): readonly QueryTerm[] {
  return [...terms]
    .map((term) => ({
      text: term.text.trim(),
      termClass: term.termClass,
      ...(term.researchOnlyOffensive ? { researchOnlyOffensive: true } : {}),
      ...(term.sourceId ? { sourceId: term.sourceId.trim() } : {}),
      ...(term.weight !== undefined ? { weight: term.weight } : {}),
    }))
    .sort((left, right) => {
      const byClass = left.termClass.localeCompare(right.termClass);
      if (byClass !== 0) {
        return byClass;
      }
      return left.text.localeCompare(right.text);
    });
}

function canonicalizePackContent(input: CanonicalPackContent): string {
  const payload = {
    id: input.id,
    displayName: input.displayName.trim(),
    entityKind: input.entityKind,
    theme: input.theme,
    semver: input.semver.trim(),
    terms: canonicalizeTerms(input.terms),
    ...(input.notes ? { notes: input.notes.trim() } : {}),
  };
  return JSON.stringify(payload);
}

export function computeQueryPackContentHash(input: CanonicalPackContent): string {
  return hashUtf8(canonicalizePackContent(input)).digest;
}

export function buildQueryPackVersionId(semver: string, contentHash: string): QueryPackVersionId {
  const shortHash = contentHash.slice(0, 8);
  return `${semver}+${shortHash}` as QueryPackVersionId;
}

export function buildQueryPackVersion(semver: string, contentHash: string): QueryPackVersion {
  assertSemverValid(semver);
  return { semver: semver.trim(), contentHash };
}

export type BuildQueryPackInput = {
  readonly id: string;
  readonly displayName: string;
  readonly entityKind: EntityKind;
  readonly theme: QueryPackTheme;
  readonly semver: string;
  readonly terms: readonly QueryTerm[];
  readonly createdAt: string;
  readonly notes?: string;
};

export function buildQueryPack(input: BuildQueryPackInput): QueryPack {
  assertSemverValid(input.semver);
  assertQueryTermsValid(input.terms);
  if (!input.id.trim()) {
    throw new Error('Query pack id is required');
  }
  if (!input.displayName.trim()) {
    throw new Error('Query pack displayName is required');
  }

  const contentHash = computeQueryPackContentHash({
    id: input.id,
    displayName: input.displayName,
    entityKind: input.entityKind,
    theme: input.theme,
    semver: input.semver,
    terms: input.terms,
    ...(input.notes !== undefined ? { notes: input.notes } : {}),
  });
  const version = buildQueryPackVersion(input.semver, contentHash);
  const versionId = buildQueryPackVersionId(version.semver, version.contentHash);

  return {
    schemaVersion: QUERY_PACK_SCHEMA_VERSION,
    id: input.id.trim(),
    displayName: input.displayName.trim(),
    entityKind: input.entityKind,
    theme: input.theme,
    version,
    versionId,
    terms: input.terms,
    createdAt: input.createdAt,
    ...(input.notes !== undefined ? { notes: input.notes.trim() } : {}),
  };
}

export function assertQueryPackValid(pack: QueryPack): void {
  if (pack.schemaVersion !== QUERY_PACK_SCHEMA_VERSION) {
    throw new Error(
      `Query pack schema version mismatch: expected ${QUERY_PACK_SCHEMA_VERSION}, got ${pack.schemaVersion}`,
    );
  }
  assertQueryTermsValid(pack.terms);
  const expectedHash = computeQueryPackContentHash({
    id: pack.id,
    displayName: pack.displayName,
    entityKind: pack.entityKind,
    theme: pack.theme,
    semver: pack.version.semver,
    terms: pack.terms,
    ...(pack.notes !== undefined ? { notes: pack.notes } : {}),
  });
  if (pack.version.contentHash !== expectedHash) {
    throw new Error('Query pack contentHash does not match canonical content');
  }
  const expectedVersionId = buildQueryPackVersionId(pack.version.semver, pack.version.contentHash);
  if (pack.versionId !== expectedVersionId) {
    throw new Error('Query pack versionId does not match semver+contentHash');
  }
}

type RawQueryPackFixture = {
  readonly schemaVersion: string;
  readonly pack: {
    readonly id: string;
    readonly displayName: string;
    readonly entityKind: EntityKind;
    readonly theme: QueryPackTheme;
    readonly version: { readonly semver: string };
    readonly terms: readonly QueryTerm[];
    readonly createdAt: string;
    readonly notes?: string;
  };
  readonly expectations: QueryPackFixture['expectations'];
};

export function parseQueryPackFixture(raw: unknown): QueryPackFixture {
  if (typeof raw !== 'object' || raw === null) {
    throw new Error('Query pack fixture must be an object');
  }
  const fixture = raw as RawQueryPackFixture;
  if (fixture.schemaVersion !== QUERY_PACK_SCHEMA_VERSION) {
    throw new Error(`Unsupported fixture schema: ${String(fixture.schemaVersion)}`);
  }
  const pack = buildQueryPack({
    id: fixture.pack.id,
    displayName: fixture.pack.displayName,
    entityKind: fixture.pack.entityKind,
    theme: fixture.pack.theme,
    semver: fixture.pack.version.semver,
    terms: fixture.pack.terms,
    createdAt: fixture.pack.createdAt,
    ...(fixture.pack.notes !== undefined ? { notes: fixture.pack.notes } : {}),
  });
  assertQueryPackValid(pack);
  return { schemaVersion: QUERY_PACK_SCHEMA_VERSION, pack, expectations: fixture.expectations };
}

export function evaluateTextAgainstTerms(
  text: string,
  terms: readonly QueryTerm[],
): readonly QueryTerm[] {
  const normalized = text.toLowerCase();
  return terms.filter((term) => normalized.includes(term.text.toLowerCase()));
}

File: workers/research/src/black_book_research/query_packs/pack.py
"""Query pack construction, canonical hashing, and version identity (BB-038)."""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass

from .terms import assert_query_terms_valid
from .types import EntityKind, QueryPack, QueryPackTheme, QueryPackVersion, QueryTerm

QUERY_PACK_SCHEMA_VERSION = "query-pack.v1"
_SEMVER = re.compile(r"^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$")


def assert_semver_valid(semver: str) -> None:
    if not _SEMVER.match(semver.strip()):
        raise ValueError(f"Invalid semver for query pack: {semver}")


def _term_payload(term: QueryTerm) -> dict[str, object]:
    payload: dict[str, object] = {
        "text": term.text.strip(),
        "termClass": term.term_class,
    }
    if term.research_only_offensive:
        payload["researchOnlyOffensive"] = True
    if term.source_id:
        payload["sourceId"] = term.source_id.strip()
    if term.weight is not None:
        payload["weight"] = term.weight
    return payload


def _canonicalize_terms(terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> list[dict[str, object]]:
    return sorted((_term_payload(term) for term in terms), key=lambda item: (item["termClass"], item["text"]))


def compute_query_pack_content_hash(
    *,
    id: str,
    display_name: str,
    entity_kind: EntityKind,
    theme: QueryPackTheme,
    semver: str,
    terms: tuple[QueryTerm, ...] | list[QueryTerm],
    notes: str | None = None,
) -> str:
    payload: dict[str, object] = {
        "id": id,
        "displayName": display_name.strip(),
        "entityKind": entity_kind,
        "theme": theme,
        "semver": semver.strip(),
        "terms": _canonicalize_terms(terms),
    }
    if notes:
        payload["notes"] = notes.strip()
    canonical = json.dumps(payload, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def build_query_pack_version_id(semver: str, content_hash: str) -> str:
    return f"{semver.strip()}+{content_hash[:8]}"


@dataclass(frozen=True, slots=True)
class BuildQueryPackInput:
    id: str
    display_name: str
    entity_kind: EntityKind
    theme: QueryPackTheme
    semver: str
    terms: tuple[QueryTerm, ...] | list[QueryTerm]
    created_at: str
    notes: str | None = None


def build_query_pack(input: BuildQueryPackInput) -> QueryPack:
    assert_semver_valid(input.semver)
    terms_tuple = tuple(input.terms)
    assert_query_terms_valid(terms_tuple)
    if not input.id.strip():
        raise ValueError("Query pack id is required")
    if not input.display_name.strip():
        raise ValueError("Query pack displayName is required")

    content_hash = compute_query_pack_content_hash(
        id=input.id,
        display_name=input.display_name,
        entity_kind=input.entity_kind,
        theme=input.theme,
        semver=input.semver,
        terms=terms_tuple,
        notes=input.notes,
    )
    version = QueryPackVersion(semver=input.semver.strip(), content_hash=content_hash)
    version_id = build_query_pack_version_id(version.semver, version.content_hash)

    return QueryPack(
        schema_version=QUERY_PACK_SCHEMA_VERSION,
        id=input.id.strip(),
        display_name=input.display_name.strip(),
        entity_kind=input.entity_kind,
        theme=input.theme,
        version=version,
        version_id=version_id,
        terms=terms_tuple,
        created_at=input.created_at,
        notes=input.notes.strip() if input.notes else None,
    )


def assert_query_pack_valid(pack: QueryPack) -> None:
    if pack.schema_version != QUERY_PACK_SCHEMA_VERSION:
        raise ValueError(
            f"Query pack schema version mismatch: expected {QUERY_PACK_SCHEMA_VERSION}, "
            f"got {pack.schema_version}"
        )
    assert_query_terms_valid(pack.terms)
    expected_hash = compute_query_pack_content_hash(
        id=pack.id,
        display_name=pack.display_name,
        entity_kind=pack.entity_kind,
        theme=pack.theme,
        semver=pack.version.semver,
        terms=pack.terms,
        notes=pack.notes,
    )
    if pack.version.content_hash != expected_hash:
        raise ValueError("Query pack content_hash does not match canonical content")
    expected_version_id = build_query_pack_version_id(pack.version.semver, pack.version.content_hash)
    if pack.version_id != expected_version_id:
        raise ValueError("Query pack version_id does not match semver+content_hash")


def evaluate_text_against_terms(text: str, terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> tuple[QueryTerm, ...]:
    normalized = text.lower()
    return tuple(term for term in terms if term.text.lower() in normalized)
